src/nvim/event/socket.c

```diff
@@ -90,14 +71,38 @@ int socket_watcher_start(SocketWatcher *watcher, int backlog, socket_cb cb)
   int result;
 
   if (watcher->stream->type == UV_TCP) {
-    result = uv_tcp_bind(&watcher->uv.tcp.handle,
-                         (const struct sockaddr *)&watcher->uv.tcp.addr, 0);
+    struct addrinfo *ai = watcher->uv.tcp.addrinfo;
+
+    for (; ai; ai = ai->ai_next) {
+      memcpy(&watcher->uv.tcp.addr, ai->ai_addr, ai->ai_addrlen);
+      result = uv_tcp_bind(&watcher->uv.tcp.handle,
+                           (const struct sockaddr *)&watcher->uv.tcp.addr, 0);
+      if (result != 0) {
+        continue;
+      }
+      result = uv_listen(watcher->stream, backlog, connection_cb);
+      if (result == 0) {
+        // When the endpoint in socket_watcher_init() didn't specify a port
+        // number, a free random port number will be assigned. sin_port will
+        // contain 0 in this case, unless uv_tcp_getsockname() is used first.
+        uv_tcp_getsockname(&watcher->uv.tcp.handle,
+                           (struct sockaddr *)&watcher->uv.tcp.addr,
```

Comment:
Why not use a local `struct sockaddr_storage x` here? The content of `x` is not used in this call. Member `watcher->uv.tcp.addr` could be removed.